### ISSUE TITLE
Saving/Caching for offline use?

### ISSUE THREAD

**Issue by bpevs** [protomaps/PMTiles#395]
I am currently happily using protomaps with maplibregl for maps on the web!

I haven't seen any mention of offline use, so I'm wondering if this use-case has been considered at some point?  One of the benefits of the workflow of previous tile-based systems has been that I could actually download and cache the whole tile of an area, and have that usable offline.

However, for Protomaps, I'm currently just using one giant full-planet file.  Clearly an end-user wouldn't want to download that huge file!  Is the recommended solution just to make a bunch of pmtiles for  different areas, or is there a way to actually cache/extract/save smaller areas from a full-planet file?

**Comment by vinnyglennon** [comment]
@bpevs 
Hi,
https://gist.github.com/vinnyglennon/bb925d280741429bce4e9af2535140b2 I have been looking to save a pmtiles to indexeddb, using leaflet for display. I get an error when I retrieve it from indexeddb "Unimplemented type: 4 error".

**Comment by bdon** [comment]
Yes, `pmtiles extract` exists for this reason for getting small areas: https://docs.protomaps.com/guide/getting-started#_3-extract-any-area

Though what you're describing will require some storage mechanism that is platform-specific. It might be IndexedDB or service workers on browsers or it might be SQLite on mobile devices. These are outside of the scope of the encoding/decoding libraries in this repository, but it would be very useful to have a third-party library that accomplishes this for popular platforms.

**Comment by bpevs** [comment]
`pmtiles extract` is not meant to be run by the consumer app, though, right? So I'll probably need to do extraction separately from the app, make those downloadable, and then load those multiple downloaded pmtiles in the maplibre instance?

**Comment by bdon** [comment]
There isn't a technical constraint here, the consumer could extract a region of tiles like `pmtiles extract`, but I don't believe an implementation of the extract algorithm exists outside of the Go one.

**Comment by bpevs** [comment]
Cool got it!  Thanks for the info!  I'll make a discussion instead of an issue next time, now that I realize it's there :+1:

**Comment by bpevs** [comment]
fwiw if anyone finds this, I ended up using the more-standard pattern of creating pmtiles extracts that users can download. It does work as an offline tauri app, though! I have a smaller example repo:

https://github.com/inro-digital/tauri-offline-maps

**Comment by jtbaker** [comment]
> fwiw if anyone finds this, I ended up using the more-standard pattern of creating pmtiles extracts that users can download. It does work as an offline tauri app, though! I have a smaller example repo:
> 
> https://github.com/inro-digital/tauri-offline-maps

Cool @bpevs, I'm working on implementing a similar pattern with a Tauri app. I noticed that you seem to be serving the tiles from a local deno webserver in the app? I'm going to see about saving it to indexeddb as a blob and then accessing it using the `Blob.slice(start, end).arrayBuffer()` pattern to hopefully make things a little more efficient.

**Comment by bpevs** [comment]
Mmm not a web server, but tauri fs methods.  I had figured blobs would be too big for something like indexdb (at least for my purposes).  But I guess deps on map size.

**Comment by bpevs** [comment]
> Cool @bpevs, I'm working on implementing a similar pattern with a Tauri app. I noticed that you seem to be serving the tiles from a local deno webserver in the app? I'm going to see about saving it to indexeddb as a blob and then accessing it using the `Blob.slice(start, end).arrayBuffer()` pattern to hopefully make things a little more efficient.

@jtbaker did you ever get this impl working? Lately, I've been interested offline maps for PWA (rather than Tauri), and so in that case I would need to use indexdb. So if you haven't, maybe I'll try my hand at that.

**Comment by bdon** [comment]
> Do you think an implementation of this would be in scope for this core repo?

I'd prefer to keep the footprint of the library small like it is currently. 

> several utility methods/types from the lib I would like to use in the implementation, that weren't exported as part of the public API I would need to vendor.

Can you give a few examples? Maybe we ought to expose those.

**Comment by jtbaker** [comment]
> Can you give a few examples? Maybe we ought to expose those. 

I'll work on getting my POC fleshed out and then bring a more complete proposal on what would be nice to have. Been a few months since I was deep into it. Thank you!

**Comment by jtbaker** [comment]
@bdon thanks for the pointer - a little 🤦 on my part, but I appreciate you taking the time to look at the code and noticing that omission - it got me unstuck!

@bpevs check out this package I just created - https://www.npmjs.com/package/pmtiles-offline, source at https://github.com/jtbaker/pmtiles-offline in case it's useful. I'm definitely interested in feedback about the API/interface and open to feedback/contributions.

<summary>

<span>I'm seeing pretty decent performance according to some debug logs I added for the retrieval mechanism - usually in the neighborhood of ~1ms for tile retrieval, even with IndexedDB's transaction overhead</span>

<details>

```
[IndexedDB] Starting retrieval: range=385566-415427 (29861 bytes)
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=385566-415427 (29861 bytes) in 0.50ms
index.ts:76 [IndexedDB] Starting retrieval: range=415427-461961 (46534 bytes)
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=415427-461961 (46534 bytes) in 1.60ms
index.ts:76 [IndexedDB] Starting retrieval: range=461961-516943 (54982 bytes)
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=461961-516943 (54982 bytes) in 0.40ms
index.ts:76 [IndexedDB] Starting retrieval: range=516943-547833 (30890 bytes)
index.ts:76 [IndexedDB] Starting retrieval: range=731432-811687 (80255 bytes)
index.ts:76 [IndexedDB] Starting retrieval: range=547833-629569 (81736 bytes)
index.ts:76 [IndexedDB] Starting retrieval: range=629569-731432 (101863 bytes)
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=516943-547833 (30890 bytes) in 0.60ms
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=731432-811687 (80255 bytes) in 0.80ms
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=547833-629569 (81736 bytes) in 1.10ms
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=629569-731432 (101863 bytes) in 1.30ms
index.ts:76 [IndexedDB] Starting retrieval: range=811687-834533 (22846 bytes)
index.ts:76 [IndexedDB] Starting retrieval: range=1177251-1226224 (48973 bytes)
index.ts:76 [IndexedDB] Starting retrieval: range=897827-990932 (93105 bytes)
index.ts:76 [IndexedDB] Starting retrieval: range=990932-1110558 (119626 bytes)
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=811687-834533 (22846 bytes) in 0.60ms
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=1177251-1226224 (48973 bytes) in 0.70ms
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=897827-990932 (93105 bytes) in 0.80ms
index.ts:91 [IndexedDB] Retrieved and sliced blob: range=990932-1110558 (119626 bytes) in 0.90ms
```

</details>
</summary>

I also had Claude Code (🦆 ) put together a quick demo PWA page showing the functionality: https://jtbaker.github.io/pmtiles-offline/

**Comment by bpevs** [comment]
@jtbaker Sorry it took me so long to get around to it, but your package was easy to use, and works perfectly! Really awesome work! Maybe adding a `--quiet` option to avoid informational log spam would be my only qualm. Awkwardly, I'm not currently working on the project I originally wanted to use offline tiles for, but I've been stuck at home sick the last few days, and found some time to make a thing (errr... I say a few days, but mostly it was piecing together prior things I worked on, so it's probably more than a few days work heh)!

Built https://maps.bpev.me, which is an offline-capable pwa maps app for saving bookmarks (I guess like CoMaps/OrganicMaps sans navigation). Still need to sort out planet-wide/zoom-transition behavior, some usability, and performance things (I mean performance issues on my side; haven't optimized too much)... but the "download regional tile -> use it offline" path works great! Thanks both of you for all your work on pmtiles and pmtiles-offline!

https://codeberg.org/bpev/maps-app

<img width="4380" height="2420" alt="Image" src="https://github.com/user-attachments/assets/44bd1920-bda2-40a2-a569-f98d11ba425d" />

<img width="720" height="620" alt="Image" src="https://github.com/user-attachments/assets/19b092eb-4eb8-408c-b445-3c54662f3ae5" />

**Comment by jtbaker** [comment]
@bpevs that makes me so happy, I'm glad you found it useful! Looks like a neat project - I like having a more opinionated general purpose navigation focused version in addition to my more generic "x-ray" version that will render any generic PMTiles file (now that I think about it, I should probably enable rendering for PMTiles rasters too).

 If you're up for it would happily accept a PR to update the behavior around the logging.
